drivers/adc/adc_stm32.c

```diff
@@ -642,7 +642,8 @@ static const uint32_t table_oversampling_ratio[] = {
 static void adc_stm32_oversampling_scope(ADC_TypeDef *adc, uint32_t ovs_scope)
 {
 #if defined(CONFIG_SOC_SERIES_STM32L0X) || \
-	defined(CONFIG_SOC_SERIES_STM32WLX)
+	defined(CONFIG_SOC_SERIES_STM32WLX) || \
+	defined(CONFIG_SOC_SERIES_STM32G0X)
```

Comment:
STM32G0 doesn't act the exact same way as L0 or WL.
For G0 (from Reference Manual RM0454) "Software is allowed to write this bit only when ADSTART = 0 (which ensures that no conversion is ongoing)."
For L0 (RM0451) and WL (RM0453) "The software is allowed to write this bit only when ADEN bit is cleared."

So for L0/WL, ADC needs to be disabled, for most other series, it just needs to have no ongoing conversion.
I'd rather avoid disabling the ADC if not strictly necessary as it has been showed in another PR to cause performance issues (due to ADC enable taking some time).

So I'd prefer a change like this:
```
	if (LL_ADC_GetOverSamplingScope(adc) == ovs_scope) {
		return;
	}
#if defined(CONFIG_SOC_SERIES_STM32L0X) || \
	defined(CONFIG_SOC_SERIES_STM32WLX)
	/* ADC must be disabled before setting OVS bits */
	adc_stm32_disable(adc);
#else
	/* ADC must have no ongoing conversion before setting OVS bits */
	if (LL_ADC_REG_IsConversionOngoing(adc)) {
		LL_ADC_REG_StopConversion(adc);
		while (LL_ADC_REG_IsConversionOngoing(adc)) {
		}
	}
#endif
	LL_ADC_SetOverSamplingScope(adc, ovs_scope);

```

As an aside, if you're currently facing an issue, it probably means that you have an ongoing conversion when you configure the oversampling. With your solution or mine, your conversion will be stopped and lost. Might be worth to look into it.

Reply:
Hello, sorry if I wasn't specific enough.
My current issue isn't that I am not able to write the oversampling bits, it is that when writing to the ADC_CFRG2 bit while ADEN is active, another bit (CKMODE) is set back to 0. 
I've reproduced the issue on a fresh clone of Zephyr : 
1. In /zephyr/samples/drivers/board create a file named nucleo_g0b1re.overlay containing :
```
/ {
	zephyr,user {
		io-channels = <&adc1 0>, <&adc1 1>;
	};
};

&adc1 {
	#address-cells = <1>;
	#size-cells = <0>;
 
	channel@0 {
		reg = <0>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_INTERNAL";
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,resolution = <12>;
	};
	
	channel@1 {
		reg = <1>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_INTERNAL";
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,resolution = <12>;
	};
};
```
2. build and flash the example code for the adc in samples/drivers/adc
3. start debugging and set a breakpoint on the line err = adc_stm32_oversampling(adc, sequence->oversampling); (around line 970 in adc_stm32.c)
4. Check the state of ADEN, ADSTART and CKMODE before and after breaking

I've included some screenshots with my reply, showing the states of the registers before and after the oversampling write.

![before_OVSE_set](https://github.com/zephyrproject-rtos/zephyr/assets/162132562/7b39d623-dc82-480b-b499-a2f4590f9299)
![after_OVSE_set](https://github.com/zephyrproject-rtos/zephyr/assets/162132562/e56dff8b-7aba-4372-be47-566740fc5f06)